### PR TITLE
19 user registration request

### PR DESCRIPTION
**Pull request by Mateusz-Dobrzynski**
**Review this pull request with caution, as it will serve as a template for future request handlers.**
~~This PR is marked as a draft until @aleksanderhanski ensures that all tests of the `emailCheck()` method are green ✅.~~

### CODE REVIEW

**Review by aleksanderhanski — APPROVED** [review]

File: src/main/java/org/lookout_studios/meals_management_system/meals_management_system/DatabaseService.java
Comment (hubertborysowski) [inline src/main/java/org/lookout_studios/meals_management_system/meals_management_system/DatabaseService.java:1]: Line 81: typo in "*from*" (I can't comment it directly).

File: src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Checks if a given string is a valid email address
+     * 
+     * @param email An email address to be checked
+     * @return True if a string is a valid email address, false if it isn't
+     */
+    public boolean emailCheck(String email) {
+        boolean match;
+        try {
+            Pattern pattern = Pattern.compile("^[\\w+-]+(\\.[\\w+-]+)*[\\.]?[a-zA-Z0-9]@([a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$");
```
Comment (hubertborysowski) [inline src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java]: Magic value.

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            @Override
+            public void handleError(ClientHttpResponse response) throws IOException {
+            }
+        });
+    }
+
+    @AfterAll
+    public static void afterAll() {
+        wireMockServer.stop();
+    }
```
Comment (hubertborysowski) [inline src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java:49]: Error while running tests.
```java
com.github.tomakehurst.wiremock.common.FatalStartupException: java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:8080
 at com.github.tomakehurst.wiremock.WireMockServer.start([WireMockServer.java:160]
```
2023-08-19 18_03_56-RegistrationServiceTests java - meals-management-system - Visual Studio Code

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    @Test
+    public void registerExistingUserFail() throws Exception {
+        JSONObject requestBody = new JSONObject();
+        requestBody.put("email", "abc@mail.com");
+        requestBody.put("password", "12345678");
+        String responseBody = new ResponseBody(HttpStatus.OK).getResponseBody();
+        WireMock.stubFor(
+                WireMock.post(WireMock.urlEqualTo("/register"))
+                        .willReturn(WireMock.aResponse().withStatus(200).withBody(responseBody)));
```
Comment (hubertborysowski) [inline src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java]: Magic value (200).

**Review by hubertborysowski — CHANGES_REQUESTED** [review]
1 Error has occurred during tests. Some typos and magic values has been found.

Also, I'm not sure whether I should get some response or some kind of a message when I enter localhost:8080 or localhost:8080/register, or not, as I didn't get any.

File: src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java (hunk 1 above)
Comment (Mateusz-Dobrzynski) [reply src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java]: @aleksanderhanski, please resolve this and avoid using magic values in future commits

**Comment by Mateusz-Dobrzynski** [comment]
@hubertborysowski, paths of requests, e.g. `/register` are not meant to be opened in a browser. They are meant to receive HTTP requests sent by other applications. You can test the API using Postman. Please note that you need to install Postman Agent and deploy the database, if you want to test a locally hosted API.

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        requestBody.put("email", "abc-@mail.com");
+        requestBody.put("password", "12345678");
+        String responseBody = new ResponseBody(HttpStatus.BAD_REQUEST).getResponseBody();
+        WireMock.stubFor(
+                WireMock.post(WireMock.urlEqualTo("/register"))
+                        .willReturn(WireMock.aResponse().withStatus(400).withBody(responseBody)));
+                        /**
+                         * response with status 400 Bad Request 
+                         * the server will not process the request due to a client error
+                         */
```
Comment (Mateusz-Dobrzynski) [inline src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java]: @aleksanderhanski, if possible, magic values should be removed by improving code readability, not by explaining its unclear parts with comments. You can remove these comments and e.g. use `HttpStatus.BAD_REQUEST.value()` in the source code.

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        requestBody.put("email", "abc@mail.com");
+        requestBody.put("password", null);
+        String responseBody = new ResponseBody(HttpStatus.BAD_REQUEST).getResponseBody();
+        WireMock.stubFor(
+                WireMock.post(WireMock.urlEqualTo("/register"))
+                        .willReturn(WireMock.aResponse().withStatus(400).withBody(responseBody)));
+                        /**
+                         * response with status 400 Bad Request 
+                         * the server will not process the request due to a client error
+                         */
```
Comment (Mateusz-Dobrzynski) [inline src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java]: Unnessecary comments, refer to [this comment](https://github.com/Mateusz-Dobrzynski/meals-management-system/pull/9#discussion_r1300297585)

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java (hunk 3 above)
Comment (Mateusz-Dobrzynski) [reply src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java]: Refer to [this comment](https://github.com/Mateusz-Dobrzynski/meals-management-system/pull/9#discussion_r1300297585)

File: src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        JSONObject requestBody = new JSONObject();
+        requestBody.put("email", "abc@mail.com");
+        requestBody.put("password", "12345678");
+        String responseBody = new ResponseBody(HttpStatus.OK).getResponseBody();
+        WireMock.stubFor(
+                WireMock.post(WireMock.urlEqualTo("/register"))
+                        .willReturn(WireMock.aResponse().withStatus(200).withBody(responseBody)));
+                        /**
+                         * response with status 200 OK - the request succeeded
+                         */
```
Comment (Mateusz-Dobrzynski) [inline src/test/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationServiceTests.java]: Unnessecary comments, refer to [this comment](https://github.com/Mateusz-Dobrzynski/meals-management-system/pull/9#discussion_r1300297585)

**Review by Mateusz-Dobrzynski — COMMENTED** [review]
@aleksanderhanski, in some cases magic were explained with comments instead of being removed. Please address this issue

File: src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Checks if a given string is a valid email address
+     * 
+     * @param email An email address to be checked
+     * @return True if a string is a valid email address, false if it isn't
+     */
+    public boolean emailCheck(String email) {
+        boolean match;
+        try {
+            Pattern validEmailPattern = Pattern.compile("^[\\w+-]+(\\.[\\w+-]+)*[\\.]?[a-zA-Z0-9]@([a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$");
```
Comment (hubertborysowski) [inline src/main/java/org/lookout_studios/meals_management_system/meals_management_system/RegistrationService.java]: Magic value is still present. Try defining it at the beginning of the class as static string with descriptive variable name and put it down here (as a variable).

**Review by hubertborysowski — CHANGES_REQUESTED** [review]
One magic value is still present. @aleksanderhanski make sure you refer to what I said in the comment on github and to our conversation on messenger as well.

**Review by hubertborysowski — APPROVED** [review]
Approved. Couldn't find anything else against merging.

**Comment by Mateusz-Dobrzynski** [comment]
@hubertborysowski, does the error you previously found still occur? I'm currently unable to check it myself.

**Comment by hubertborysowski** [comment]
> @hubertborysowski, does the error you previously found still occur? I'm currently unable to check it myself.

Well, I discovered that this error is caused only when I want to run unit test while the app is running. Error is probably caused by the fact that the unit test wants to take the same port as the main app. Once the port is free everything works fine.

Edit: If it shouldn't be that way, I'll change my review then.

**Comment by Mateusz-Dobrzynski** [comment]
>Error is probably caused by the fact that the unit test wants to take the same port as the main app.

Yes, this is probably the reason, especially that this test was green when I created the pull request.

**Pull request merged**